### PR TITLE
Better store management

### PR DESCRIPTION
**Pull request by magnatelee**
This PR contains changes to eagerly collect stores that are unused throughout the execution. The PR also makes two small improvements for better field reuse: 1) allowing fields to be reused for any stores with the same size; 2) allowing the region field of a temporary store to be "moved" to another store within the same operation. Finally, the PR also eliminates costly consensus matches for single node execution.

### CODE REVIEW

File: legate/core/runtime.py [hunk 1]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
+            self.runtime.destroy_region_manager(
+                self.shape, region, unordered=not ordered
+            )
+
+    def remove_all_fields(self, region: Region) -> None:
+        free_fields: Deque[tuple[Region, int]] = deque()
+        for pair in self.free_fields:
+            if pair[0] != region:
+                free_fields.append(pair)
+        self.free_fields = free_fields
```
Comment (bryevdv) [inline legate/core/runtime.py:319]: can simplify and avoid the explicit type (I think) by using a generator:
```suggestion
        self.free_fields = deque(f for f in self.free_fields if f[0] != region)
```

File: legate/core/runtime.py [hunk 2]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
 
-        region_manager = self.runtime.find_or_create_region_manager(self.shape)
-        return region_manager.allocate_field(self.dtype)
+        self._field_match_manager.update_free_fields()
+
+        # Check again to see if we have any free fields
+        if len(self.free_fields) > 0:
+            return self.free_fields.popleft()
+
+        return None
```
Comment (bryevdv) [inline legate/core/runtime.py:342]: consider ternary

File: legate/core/runtime.py [hunk 3]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
+        region_manager = self.runtime.find_or_create_region_manager(self.shape)
+        return region_manager.allocate_field(self.field_size)
+
+    def free_field(
+        self, region: Region, field_id: int, ordered: bool = False
+    ) -> None:
+        self.free_fields.append((region, field_id))
+        region_manager = self.runtime.find_region_manager(region)
+        destroy = region_manager.decrease_field_count()
+        if destroy:
```
Comment (bryevdv) [inline legate/core/runtime.py:309]: consider walrus

File: legate/core/solver.py [hunk 4]
```diff
@@ -173,6 +181,12 @@ def get_field_space(self, part: PartSym) -> FieldSpace:
     def is_key_part(self, part: PartSym) -> bool:
         return part in self._key_parts
 
+    def unify_key_part(self, part1: PartSym, part2: PartSym) -> None:
+        non_members = {part1, part2} - self._key_parts
+        if len(non_members) == 2:
+            return
+        self._key_parts.update(non_members)
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/solver.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py [hunk 5]
```diff
@@ -71,25 +72,37 @@
...
-    ) -> None:
-        self.manager = manager
+    def __init__(self, fields: List[tuple[FieldManager, Region, int]]) -> None:
         self.fields = fields
         # Allocate arrays of ints that are twice as long as fields since
         # our values will be 'field_id,tree_id' pairs
-        if len(fields) > 0:
-            alloc_string = "int[" + str(2 * len(fields)) + "]"
+        num_fields = len(fields)
+        if num_fields > 0:
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py [hunk 6]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
                     # Not found so put it back int the unordered queue
-                    self.manager.free_field(region, field_id, ordered=False)
+                    manager.free_field(region, field_id, ordered=False)
             # Notice that we do this in the order of the list which is the
             # same order as they were in the array returned by the match
-            for pair in ordered_fields:
-                if pair is None:
+            for tpl in ordered_fields:
+                if tpl is None:
                     continue
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py [hunk 7]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
         if num_fields > 0:
             # Put all the returned fields onto the ordered queue in the order
             # that they are in this list since we know
-            ordered_fields: List[Optional[tuple[Region, int]]] = [
-                None
-            ] * num_fields
-            for region, field_id in self.fields:
+            ordered_fields: List[
+                Optional[tuple[FieldManager, Region, int]]
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py [hunk 8]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
 
     def destroy(self) -> None:
         self.free_fields = deque()
-        self.freed_fields = []
+
+    def try_reuse_field(self) -> Optional[tuple[Region, int]]:
+        if len(self.free_fields) > 0:
+            return self.free_fields.popleft()
+        else:
+            return None
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py [hunk 9]
```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
...
-            self.freed_fields = list()
-            match = FieldMatch(self, local_freed_fields)
-            # Dispatch the match
-            self.runtime.dispatch(match)
-            # Put it on the deque of outstanding matches
-            self.matches.append(match)
-            # Reset the match counter back to 0
-            self.match_counter = 0
+        result = self.try_reuse_field()
+        if result is not None:
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py (hunk 1 above)
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py (hunk 2 above)
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

File: legate/core/runtime.py (hunk 3 above)
Comment (magnatelee) [reply legate/core/runtime.py]: https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46

**Comment by magnatelee** [comment]
@bryevdv thanks for the review. I think I addressed all of your comments, but let me know if I missed any or you happened to have more.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
LGTM!

**Pull request merged**